src/main/java/io/cdap/plugin/WindowAggregation.java

```diff
@@ -0,0 +1,450 @@
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    Schema inputSchema = stageConfigurer.getInputSchema();
+
+    List<String> partitionFields = config.getPartitionFields();
+    List<WindowAggregationConfig.FunctionInfo> aggregates = config.getAggregates();
+    List<String> partitionOrderFields = config.getPartitionOrderFields();
+
+    if (inputSchema == null || partitionFields.isEmpty() || aggregates.isEmpty()) {
+      stageConfigurer.setOutputSchema(null);
+      return;
+    }
+
+    validate(inputSchema, partitionFields, aggregates, partitionOrderFields, stageConfigurer.getFailureCollector());
+    stageConfigurer.getFailureCollector().getOrThrowException();
+    stageConfigurer.setOutputSchema(getOutputSchema(inputSchema, aggregates));
+  }
+
+  public void validate(Schema inputSchema, List<String> partitionFields,
+                       List<WindowAggregationConfig.FunctionInfo> aggregates, List<String> partitionOrderFields,
+                       FailureCollector collector) {
+
+    for (String partitionField : partitionFields) {
+      Schema.Field field = inputSchema.getField(partitionField);
+      if (field == null) {
+        collector.addFailure(
+          String.format("Can not partition by field '%s' because it does not exist in input schema.", partitionField),
+          String.format("Partition field '%s' must exist in input schema.", partitionField))
+          .withConfigElement("partitionFields", partitionField);
+      }
+    }
+
+    for (WindowAggregationConfig.FunctionInfo functionInfo : aggregates) {
+      String functionFieldName = functionInfo.getFieldName();
+      if (Strings.isNullOrEmpty(functionFieldName)) {
+        continue; //function will operate over partitions fields
+      }
+
+      Schema.Field inputField = inputSchema.getField(functionFieldName);
+      Schema inputFieldSchema = null;
+      if (inputField == null) {
+        collector.addFailure(
+          String.format("Invalid aggregate %s: Field '%s' does not exist in input schema.", functionInfo.description(),
+                        functionFieldName), String.format("Field '%s' must exist in input schema.", functionFieldName))
+          .withConfigProperty("aggregates");
+      } else {
+        inputFieldSchema = inputField.getSchema();
+      }
+
+      WindowAggregationConfig.Function function = functionInfo.getFunction();
+      //check input schema
+      Schema allowedInputScheme = function.getAllowedInputScheme();
+
+
+      if (allowedInputScheme != null && inputFieldSchema != null) {
+        Schema.Type type = inputFieldSchema.isNullable() ? inputFieldSchema.getUnionSchema(0).getType() :
+          inputFieldSchema.getType();
+        int indexOf = allowedInputScheme.getUnionSchemas().stream().map(Schema::getType).collect(Collectors.toList())
```

Comment:
why are we getting union schemas? would these always be union input schemas?

Reply:
Yes, this will  always be union schema, because we defined it.
Reference: https://github.com/data-integrations/window-aggregation/blob/f713c67adcb557ca005d9a3f0a27919a6a07ae7e/src/main/java/io/cdap/plugin/WindowAggregationConfig.java#L309
		https://github.com/data-integrations/window-aggregation/blob/f713c67adcb557ca005d9a3f0a27919a6a07ae7e/src/main/java/io/cdap/plugin/WindowAggregationConfig.java#L314

Reply:
@CuriousVini PR updated.